SnowplowTracker/SnowplowTracker/Tracker.cs

```diff
@@ -0,0 +1,342 @@
+using SnowplowTracker.Events;
+using SnowplowTracker.Storage;
+using SnowplowTracker.Enums;
+
+namespace SnowplowTracker {
+	public class Tracker {
+
+		private Emitter emitter;
+		private Subject subject;
+		private Session session;
+		private string trackerNamespace;
+		private string appId;
+		private DevicePlatforms platform;
+		private bool base64Encoded;
+
+		private bool dataCollection = true;
+		private bool synchronous = false;
+	
+		/// <summary>
+		/// Initializes a new instance of the <see cref="SnowplowTracker.Tracker"/> class.
+		/// </summary>
+		/// <param name="emitter">Emitter used for all sending and storing operations.</param>
+		/// <param name="trackerNamespace">Tracker namespace.</param>
+		/// <param name="appId">App identifier.</param>
+		/// <param name="platform">The DevicePlatform the tracker is running on.</param>
+		/// <param name="base64Encoded">If set to <c>true</c> all unstructured events and contextes will be base64 encoded.</param>
+		public Tracker(Emitter emitter, Subject subject, Session session, string trackerNamespace, string appId, DevicePlatforms platform = null, 
+		               bool base64Encoded = true) {
+			// Preconditions
+			Utils.CheckArgument(emitter != null, "Emitter cannot be null.");
+
+			// Tracker Setup
+			this.emitter = emitter;
+			this.trackerNamespace = trackerNamespace;
+			this.appId = appId;
+			this.platform = (platform != null) ? platform : DevicePlatforms.Mobile;
+			this.base64Encoded = base64Encoded;
+			this.session = session;
+
+			// Boot the emitter..
+			this.emitter.Start ();
+		}
+
+		// --- Tracker Control Methods
+
+		/// <summary>
+		/// Pauses the event tracking.
+		/// </summary>
+		public void PauseEventTracking() {
+			dataCollection = false;
+			if (session != null) {
+				session.StopSessionChecker();
+			}
+			emitter.Stop();
+		}
+
+		/// <summary>
+		/// Resumes the event tracking.
+		/// </summary>
+		public void ResumeEventTracking() {
+			dataCollection = true;
+			if (session != null) {
+				session.StartSessionChecker();
+			}
+			emitter.Start();
+		}
+
+		/// <summary>
+		/// Sets the tracker to synchronous operation.
+		/// </summary>
+		public void SetSynchronous() {
+			synchronous = true;
+			this.emitter.synchronous = true;
+		}
+
+		/// <summary>
+		/// Sets the tracker to asynchronous operation.
+		/// </summary>
+		public void SetAsynchronous() {
+			synchronous = false;
+			this.emitter.synchronous = false;
+		}
+
+		// --- Event Tracking
+
+		/// <summary>
+		/// Track the specified event.
+		/// - Checks whether dataCollection is on
+		/// - Checks whether to use synchronous sending or not
+		/// </summary>
+		/// <param name="newEvent">New event to track.</param>
+		public void Track(IEvent newEvent) {
+			if (!dataCollection) {
+				return;
+			}
+			if (synchronous) {
+				ProcessEvent (newEvent);
+			} else {
+				ThreadPool.QueueUserWorkItem (new WaitCallback ( delegate(object t) {
+					ProcessEvent (newEvent);
```

Comment:
This is my understanding of what goes on prior to events getting in to the database:

Tracker.track creates a short-lived asynchronous task which turns the IEvent into a TrackerPayload, then adds it to the eventStore's concurrent queue. The eventStore thread wakes up and inserts the event into the database.

My suggestion: eliminate the creation of the asynchronous task. This would simplify things because we would just have the UI thread and the PayloadQueueConsumer thread, rather than also using ThreadPool. So either do the work of ProcessEvent in the UI thread (this is what currently happens if `synchronous` is true`) or do it on the other side of the queue.

Update: by the time I finished typing this out, I'm not so sure. The disadvantage of the first option is that tracker calls will block the UI thread for a bit longer. The disadvantage of the second option is that it may cause a bottleneck at the entrance to the database as instead of just writing to the database, the eventStore thread would also have to do the work of ProcessEvent. Maybe ProcessEvent is sufficiently lightweight to justify this. I'll leave this comment anyway in case anybody else has any thoughts on it.